code/modules/modular_computers/file_system/programs/coupon.dm

```diff
@@ -0,0 +1,115 @@
+		for(var/datum/coupon_code/coupon as anything in user_id.registered_account.redeemed_coupons)
+			var/list/coupon_data = list(
+				"goody" = initial(coupon.discounted_pack.name),
+				"discount" = coupon.discount*100,
+			)
+			if(coupon.printed)
+				data["printed_coupons"] += list(coupon_data)
+			else
+				data["redeemed_coupons"] += list(coupon_data)
+		data["valid_id"] = TRUE
+	return data
+
+/datum/computer_file/program/coupon/ui_act(action, params, datum/tgui/ui)
+	var/obj/item/card/id/user_id = computer.computer_id_slot
+	if(!(user_id?.registered_account))
+		return TRUE
+	switch(action)
+		if("redeem")
+			var/code = params["code"]
+			if(!length(code))
+				return TRUE
+			var/datum/coupon_code/coupon = SSmodular_computers.discount_coupons[code]
+			if(isnull(coupon))
+				user_id.registered_account.bank_card_talk("Invalid coupon code.", TRUE)
+				return TRUE
+			if(coupon.expires_in && coupon.expires_in < world.time)
+				user_id.registered_account.bank_card_talk("Expired coupon code.", TRUE)
+				return TRUE
+			if(coupon in user_id.registered_account.redeemed_coupons)
+				user_id.registered_account.bank_card_talk("Coupon [code] already redeemed.", TRUE)
+				return TRUE
+			coupon.copy(user_id.registered_account)
+			var/static/list/goodbye = list(
+				"Have a wonderful day.",
+				"Don't forget to print it.",
+				"Time to get shopping!",
+				"Enjoy your discount!",
+				"Congratulations!",
+				"Bye Bye~.",
+			)
+			user_id.registered_account.bank_card_talk("Coupon [code] redeemed. [pick(goodbye)]", TRUE)
+			//Well, guess you're redeeming something else too.
+			if(prob(40) && computer.used_capacity < computer.max_capacity)
+				var/datum/computer_file/warez = new()
+				warez.filename = random_string(rand(6, 12), GLOB.alphabet + GLOB.alphabet_upper + GLOB.numerals)
+				warez.filetype = pick("DAT", "XXX", "TMP", "FILE", "MNT", "MINER", "SYS", "PNG.EXE")
+				warez.size = min(rand(1, 4), computer.max_capacity - computer.used_capacity)
+				if(prob(25))
+					warez.undeletable = TRUE
+				computer.store_file(warez)
+
+/datum/computer_file/program/coupon/tap(atom/tapped_atom, mob/living/user, params)
+	if(!istype(tapped_atom, /obj/machinery/photocopier))
+		return FALSE
+	var/obj/item/card/id/user_id = computer.computer_id_slot
+	if(!(user_id?.registered_account))
+		computer.balloon_alert(user, "no bank account found!")
+		return TRUE
+	var/obj/machinery/photocopier/copier = tapped_atom
+	if(copier.check_busy(user))
+		return TRUE
+	var/num_coupons = 0
+	for(var/datum/coupon_code/coupon as anything in user_id.registered_account.redeemed_coupons)
+		if(!coupon.printed)
+			num_coupons++
+	if(!num_coupons)
+		computer.balloon_alert(user, "no coupon available!")
+		return TRUE
+	copier.do_copies(CALLBACK(src, PROC_REF(print_coupon), user_id.registered_account), user, COUPON_PAPER_USE, COUPON_TONER_USE, num_coupons)
+	return TRUE
+
+/datum/computer_file/program/coupon/proc/print_coupon(datum/bank_account/account)
```

Comment:
Modular computers store paper in them, is there a reason you didn't want to use those instead?

Reply:
Oh, for a moment I only thought that was possible for stationary consoles. Yeah, I'll do that instead.

Reply:
Thinking twice, I've changed my mind back to the original idea. The issue with having the PDA do all the job is that it'll make the gimmick of inputting a code in a UI, as well as some coupon codes expiring after a while seem all the more irrelevant.
I like the irony of having to pay 5 credits to print a coupon to save money, as well as putting the photocopier to use for something that's not strictly paperwork. And whil,e sure, pressing some "PRG_print" button on the UI of the program itself would seem more sensical, it's just too much of a nobrainer.

Tho, the criticism is welcome.